### PR TITLE
Allow AtomixCluster to be autowired everywhere

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR extracts building the `AtomixCluster` from the broker into the `dist` module. Unfortunately, because many of our testing utilities break all kinds of abstraction, and our tests still heavily rely on modifying the cluster via the `BrokerCfg`, mapping the `BrokerCfg` to the `AtomixCluster` is still done in the `broker` module. However, everything is prepared to eventually move this out to the dist module itself, such that our configuration would simply instantiate the Atomix `ClusterConfig` type, and later inject a `AtomixCluster` in the broker. See the `StandaloneGateway` as an example; there is plenty of opportunity to DRY it up once both are simply configuring `ClusterConfig` directly, and it will remove intermediate layers of builders/config.

In order to achieve that at the moment, since the network configuration relies on being initialized with the right broker base path, I also had to pull up the working directory into its own configuration, such that it can be autowired as well. This is a little more abstraction than I'd like, but it's due to how the `BrokerCfg` is still the central point for all the components, even those not specific to the broker.

The broker configuration is thus not initialized in the system context or broker anymore, but is given to the broker already initialized. This is done via the Spring configuration `BrokerConfiguration`. It uses two beans to do this, which I dislike, but I'm not sure what's the best way to do this. I generally dislike that the configuration must be initialized anyway, it's very brittle and easy to forget, so I'd rather figure a way to avoid this in the future. I would propose a follow up which removes this, and instead enforces that we create the broker config either passing the base already, or using a factory method which does this. I can also do it here, but it will increase how much changes are here.

I think in the future the cluster services should be started before the broker itself, but we can figure this out later. For now, it means that if the actuator wants to make use of them before the broker finishes starting, it will fail - this is something we can handle semi-gracefully, so I would leave it as is.

## Related issues

related to #9996

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
Super cool, since the SystemContext is in charge of initializing the configuration in tests, most tests fail to build a cluster (since we want to inject the cluster).

**Comment by npepinpe** [comment]
See #10356 which removes the use of two beans in the broker configuration and tries to minimize usages of uninitialized `BrokerCfg` objects.

File: dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.context.annotation.ScopedProxyMode;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.Profiles;
+import org.springframework.web.context.annotation.ApplicationScope;
+
+@Configuration(proxyBeanMethods = false)
+public record WorkingDirectoryConfiguration(Environment environment) {
+
+  @Bean
+  @ApplicationScope(proxyMode = ScopedProxyMode.NO)
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java:26]: :wrench: Default proxyMode = No. So may be it is unnecessary set it everywhere.

File: dist/src/main/java/io/camunda/zeebe/broker/BrokerClusterConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import io.atomix.utils.Version;
+import io.camunda.zeebe.broker.clustering.ClusterConfigFactory;
+import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
+import io.camunda.zeebe.util.VersionUtil;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.ScopedProxyMode;
+import org.springframework.web.context.annotation.ApplicationScope;
+
+@Configuration(proxyBeanMethods = false)
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/broker/BrokerClusterConfiguration.java:21]: :question: Why is setting this to false required?

**Review by deepthidevaki — APPROVED** [review]
Great :rocket: That was a lot of changes, but in the good direction. Thanks.

File: atomix/cluster/src/main/java/io/atomix/cluster/discovery/BootstrapDiscoveryProvider.java [hunk 3]
```diff
@@ -68,7 +68,7 @@ public BootstrapDiscoveryProvider(final Collection<Node> bootstrapNodes) {
                     .collect(Collectors.toList())));
   }
 
-  BootstrapDiscoveryProvider(final BootstrapDiscoveryConfig config) {
+  public BootstrapDiscoveryProvider(final BootstrapDiscoveryConfig config) {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/cluster/discovery/BootstrapDiscoveryProvider.java:71]: Oh yeah, this is a left over from previous changes where I was building each manually instead of the complete `ClusterConfig`.

File: atomix/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocol.java [hunk 4]
```diff
@@ -116,7 +116,7 @@
   private ScheduledFuture<?> probeFuture;
   private ScheduledFuture<?> syncFuture;
 
-  SwimMembershipProtocol(final SwimMembershipProtocolConfig config) {
+  public SwimMembershipProtocol(final SwimMembershipProtocolConfig config) {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocol.java:119]: Same as above. I think it's all one commit, we can just revert it.

File: dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java (hunk 1 above)
Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java:26]: Is it? I had issues where it tried to proxy it when setting the scope without it, on final classes :thinking: Let me try again

File: dist/src/main/java/io/camunda/zeebe/broker/BrokerClusterConfiguration.java (hunk 2 above)
Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/broker/BrokerClusterConfiguration.java:21]: It's not required, but it's more best practice for our code base. If it's true, then CGLIB will try to proxy the beans so it can apply aspects. This will fail when using final classes, which we do by default. To avoid others running into this I just set it usually (plus, we don't really make use of aspects, and proxying just makes debugging more cumbersome).

Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/broker/BrokerClusterConfiguration.java:21]: So if this is a true, a `@Bean` annotated method will not return a, say, `ActorScheduler`, but a CGLIB proxy of the `ActorScheduler`. Looks and behaves the same, but may be aspect-weaved. Downside is it must be extensible, I think (or at least it fails on final classes).

**Comment by npepinpe** [comment]
Yes, sorry, this was much more changes than I wanted, and I had to find a balance between fully moving everything shared between broker/gateway to dist and not creating too much work. At the same time, it highlighted that if we will do more with Spring, we may want to learn more about how it works :smile:

File: dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java (hunk 1 above)
Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java:26]: so the default `proxyMode` is `ScopedProxyMode.TARGET_CLASS` for me (when using `ApplicationScope` anyway).

I talked with Ole before about not needing this scope, since the default is the singleton scope, but I noticed while debugging that without it, the method was being called multiple time. I didn't dig too much into it, but maybe this is the wrong approach as well :thinking:

Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/broker/WorkingDirectoryConfiguration.java:26]: OK, I figured out why some beans are called multiple times :smile: Spring will eager initialize the beans, even if no one is requesting them for auto wiring. Since I also sometimes call the methods directly, this was causing problems. So the issue is, don't ever call bean methods directly :smile:
